### ISSUE TITLE
Should `avoid_literal_in_foreach` lint reconsider when the collection is nullable.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/linter#2751]
See: https://stackoverflow.com/questions/68271469/linter-rule-and-null-safety
In short, if the iterable is nullable, you can write `foo?.forEach((e) { ... })`, but need much more verbosity to turn it into a `for` loop, like:
```dart
var v = foo; // If it's an expression, not already a variable.
if (v != null) {
  for (var e in v) { 
    ... 
  }
}
```
over
```
foo?.forEach((e) {
  ...
});
```

According to the stack overflow report, the latter triggers the lint. Maybe it shouldn't.

**Comment by pq** [comment]
This is interesting.  I actually just recently hit this and begrudgingly made the required changes when `foo?.forEach` was so much cleaner.

I'd be really curious to get other people's thoughts...

/fyi @bwilkerson @munificent @natebosch @jakemac53 @eernstg

**Comment by lrhn** [comment]
One workaround suggested was `for (var e in [...?foo]) { ... }`. That sadly has a semantic which requires building the intermediate list.
Another was `for var e in foo ?? const <Foo>[]) { ... }` which can be optimized better, but you need to write the `<Foo>` type to preserve the type of `e`, which is also annoying.

Maybe we just need `for (var e in? foo) { ... }` like we have `[...? foo]`. 😁

**Comment by munificent** [comment]
Yeah, there's no great pattern for this. I'm OK with relaxing the lint to allow `foo?.forEach(...)`.

**Comment by pq** [comment]
OK!  I think we can consider this closed w/ #2752.

**Comment by natebosch** [comment]
> One workaround suggested was `for (var e in [...?foo]) { ... }`. That sadly has a semantic which requires building the intermediate list.


> Yeah, there's no great pattern for this. I'm OK with relaxing the lint to allow `foo?.forEach(...)`.

Another time it would be nice to have a syntax to generate iterables.
